src/metabase/pulse/render/body.clj

```diff
@@ -109,13 +124,26 @@
               :when remapped_from]
           [remapped_from col-idx])))
 
+(defn- ref-key-with-id?
+  "True if the `settings-key` is a ref and matches the id. The key will look something like:
+  [\"ref\" [\"field-ref\" id {:maybe more}]]"
+  [settings-key id]
+  (and (= "ref" (first settings-key))
+       (= id (get-in settings-key [1 1]))))
```

Comment:
Nit:

I'm really struggling to come up with names, which might be a sign, but in theory this could be made clearer by destructuring the params. "Clearer" meaning both that there's less `get-in` juggling and also the function signature more clearly communicates what shape the expected data in.

```
[[ref-label [_ ref-id]] id]
```

or something instead of

```
[settings-key id]
```

Reply:
We shouldn't be operating on this data structure. there's a way to normalize this stuff.

```clojure
viz=> (let [viz (db/select-one-field :visualization_settings 'Card :id 176)]
        {:raw viz
         :normed (mb.viz/db->norm viz)})
{:raw {:graph.dimensions ["CREATED_AT"],
       :graph.show_values true,
       :column_settings {"[\"name\",\"sum\"]" {:number_style "scientific"}},
       :graph.metrics ["sum"]},
 :normed {:graph.dimensions ["CREATED_AT"],
          :graph.show_values true,
          :graph.metrics ["sum"],
          :metabase.shared.models.visualization-settings/column-settings {{:metabase.shared.models.visualization-settings/column-name "sum"} {:metabase.shared.models.visualization-settings/number-style "scientific"}}}}
```

Check out `metabase.query-processor.middleware.visualization-settings` middleware namespace for how this can happen automatically in queries.